src/lib/pubkey/classic_mceliece/cmce_field_ordering.cpp

```diff
@@ -0,0 +1,328 @@
+void simultaneous_composeinv(secure_vector<uint16_t>& p, secure_vector<uint16_t>& q) {
+   auto p_new = composeinv(p, q);
+   q = composeinv(q, p);
+   p = std::move(p_new);
+}
+
+/**
+ * @brief Generate control bits as in ISO 9.2.10.
+ *
+ * TODO: This function can be optimized (see Classic McEliece reference implementation)
+ */
+secure_vector<uint16_t> generate_control_bits_internal(const secure_vector<uint16_t>& pi) {
+   const auto n = pi.size();
+   BOTAN_ASSERT_NOMSG(is_power_of_2(n));
+   const size_t m = ceil_log2(n);
+
+   if(m == 1) {
+      return secure_vector<uint16_t>({pi.at(0)});
+   }
+   secure_vector<uint16_t> p(n);
+   for(size_t x = 0; x < n; ++x) {
+      p.at(x) = pi.at(x ^ 1);
+   }
+   secure_vector<uint16_t> q(n);
+   for(size_t x = 0; x < n; ++x) {
+      q.at(x) = pi.at(x) ^ 1;
+   }
+
+   secure_vector<uint16_t> range_n(n);
+   std::iota(range_n.begin(), range_n.end(), 0);
+   auto piinv = composeinv(range_n, pi);
+
+   simultaneous_composeinv(p, q);
+
+   secure_vector<uint16_t> c(n);
+   for(uint16_t x = 0; static_cast<size_t>(x) < n; ++x) {
+      c.at(x) = CMCE_CT::min(x, p.at(x));
+   }
+
+   simultaneous_composeinv(p, q);
+
+   for(size_t i = 1; i < m - 1; ++i) {
+      auto cp = composeinv(c, q);
+      simultaneous_composeinv(p, q);
+      for(size_t x = 0; x < n; ++x) {
+         c.at(x) = CMCE_CT::min(c.at(x), cp.at(x));
+      }
+   }
+
+   secure_vector<uint16_t> f(n / 2);
+   for(size_t j = 0; j < n / 2; ++j) {
+      f.at(j) = c.at(2 * j) % 2;
+   }
+
+   secure_vector<uint16_t> big_f(n);
+   for(uint16_t x = 0; size_t(x) < n; ++x) {
+      big_f.at(x) = x ^ f.at(x / 2);
+   }
+
+   auto fpi = composeinv(big_f, piinv);
+
+   secure_vector<uint16_t> l(n / 2);
+   for(size_t k = 0; k < n / 2; ++k) {
+      l.at(k) = fpi.at(2 * k) % 2;
+   }
+
+   secure_vector<uint16_t> big_l(n);
+   for(uint16_t y = 0; size_t(y) < n; ++y) {
+      big_l.at(y) = y ^ l.at(y / 2);
+   }
+
+   auto big_m = composeinv(fpi, big_l);
+
+   secure_vector<uint16_t> subm0(n / 2);
+   secure_vector<uint16_t> subm1(n / 2);
+   for(size_t j = 0; j < n / 2; ++j) {
+      subm0.at(j) = big_m.at(2 * j) / 2;
+      subm1.at(j) = big_m.at(2 * j + 1) / 2;
+   }
+
+   auto subz0 = generate_control_bits_internal(subm0);
+   auto subz1 = generate_control_bits_internal(subm1);
+
+   secure_vector<uint16_t> z(subz0.size() + subz1.size());
+   for(size_t j = 0; j < subz0.size(); ++j) {
+      z.at(2 * j) = subz0.at(j);
+      z.at(2 * j + 1) = subz1.at(j);
+   }
+
+   return concat(f, z, l);
+}
+
+CT::Choice ct_has_adjacent_duplicates(std::span<const uint32_t> vec) {
+   CT::Mask<uint32_t> mask = CT::Mask<uint32_t>::cleared();
+   for(size_t i = 0; i < vec.size() - 1; ++i) {
+      mask |= CT::Mask<uint32_t>::is_equal(vec[i], vec[i + 1]);
+   }
+   return mask.as_choice();
+}
+
+}  // anonymous namespace
+
+std::optional<Classic_McEliece_Field_Ordering> Classic_McEliece_Field_Ordering::create_field_ordering(
```

Comment:
Does the non-constant time nature of `std::optional` bring up issues here? If so we should consider using `CT::Option`, if not a comment that it's fine would be helpful.

Reply:
A `std::optional` should suffice. The field ordering creation is performed in a rejection sampling loop. I.e., if the result is a nullopt, all secrets are dismissed anyway and a new rejection sampling loop starts from the beginning.